### DISCUSSION TITLE
offset does not generate rounded shapes ?

### DISCUSSION THREAD

**Discussion by mathieu-lacage** [libfive/libfive#545] [Q&A] [answered]
Hi,

I am currently trying to reproduce simple shapes to learn how to use libfive. 

I wrote the roundify function below that should be equivalent to the builtin offset function. I was expecting that the resulting rendering would display rounded corners but studio displays a cylinder with sharp corners. I suspect that there is something obvious that I am missing about libfive so, I wonder if you could point me in the right direction to understand the output generated by libfive ?

def half_space(norm, p):
    @libfive.shape.shape
    def f(x, y, z):
        return (x-p[0])*norm[0] + (y-p[1])*norm[1] + (z-p[2])*norm[2]
    return f
   
def half_space_x(delta, sign):
    return half_space((sign, 0, 0), (delta, 0, 0))

def half_space_y(delta, sign):
    return half_space((0, sign, 0), (0, delta, 0))

def half_space_z(delta, sign):
    return half_space((0, 0, sign), (0, 0, delta))

def cylinder_z(h, r):
    pz1 = half_space_z(h/2, 1)
    pz2 = half_space_z(-h/2, -1)
    @libfive.shape.shape
    def f(x, y, z):
        d = (x*x+y*y).sqrt()
        return (d - r).max(pz1.remap(x, y, z)).max(pz2.remap(x, y, z))
    return f

def roundify(shape, r):
    @libfive.shape.shape
    def f(x, y, z):
        return shape.remap(x, y, z) - r
    return f

roundify(cylinder_z(2, 1), 0.1)

**Comment by omgitsraven** [comment, marked as answer]
A function like that will only generate a rounded shape if the distance field is exact.  You can get a formula for an exact cylinder here:  https://iquilezles.org/articles/distfunctions/

**Reply by mathieu-lacage** [reply]
Hi,

I understand now that my question was really not about libfive: thanks for pointing me in the right direction. I ended up reading a lot more than expected about what makes a distance function exact, what functions are usually exact, etc. For the record, [this](https://iquilezles.org/articles/interiordistance/) article was an eye opener to help me understand why a lot of functions are not exact.

One related question: is there a specific reason why the libfive cylinder_z function is not exact ?

**Reply by mkeeter** [reply]
There's no specific reason, other than a vague intuition that bounded functions can be faster than exact ones (although as IQ points out, exact functions may end up doing less work because they provide tighter approximations; I'm not sure if that's as true for interval arithmetic as it is for raycasting).

The historical reason is that the `libfive` standard shapes library was inherited through several generations of software (see [this blog post](https://www.mattkeeter.com/blog/2017-01-09-lineage/)), and the exact-vs-bounded distinction wasn't a consideration during early development.

**Reply by mathieu-lacage** [reply]
Hi @omgitsraven,

I apologize for this very late followup but it took me a while to ponder this topic more seriously. I believe I understand what you explained and more generally, I understand that the "simple" boolean operations defined via min/max functions are not exact either.

The following example illustrates this: the resulting intersection object cannot be offset to obtain rounded corners for the intersection boundary between two simple objects:

```
from libfive.stdlib import shapes, transforms, csg, text
import libfive.shape
import studio

s = shapes.sphere(1)
b = shapes.box_exact_centered((1.8, 1.8, 1.8))
i = csg.intersection(b, s)
csg.offset(i, 0.1)


studio.set_quality(8)
studio.set_resolution(60)
studio.set_bounds([-10, -10, -10], [10, 10, 10])
```

So, my question is: how do those who use such frep systems manage rounded corners in general? Do they use different definitions of boolean operations?

**Reply by omgitsraven** [reply]
Unfortunately it's not something I have much experience with; my impression is that it's used on exact shape definitions alone, *before* any boolean operations.  One of the fun things about SDFs is that there's lots of different ways you can transform them, but unfortunately, many of those transformations, even the simplest ones, break some of the assumptions about what it's possible to do with a "pure" SDF.
